File: java/src/main/java/myorg/common/MyByteBuffer.java
package myorg.common;

import com.google.common.hash.Hashing;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class MyByteBuffer {

    private ByteBuffer byteBuffer;

    public MyByteBuffer() {
        this(128, ByteOrder.LITTLE_ENDIAN);
    }

    public MyByteBuffer(int capacity, ByteOrder bo) {
        this.byteBuffer = ByteBuffer.allocate(capacity);
        this.byteBuffer.order(bo);
    }

    public MyByteBuffer clear() {
        byteBuffer.clear();
        return this;
    }

    public int hashCode() {
        return hashCode(0);
    }

    public int hashCode(int seed) {
        byte[] array = byteBuffer.array();
        int len = byteBuffer.position();
        
        return Hashing.murmur3_32(seed).hashBytes(array, 0, len).asInt();
    }

    public MyByteBuffer addString(String value) {
        byteBuffer.put(value.getBytes());
        return this;
    }

    public MyByteBuffer addInt(int value) {
        byteBuffer.putInt(value);
        return this;
    }

    public MyByteBuffer addLong(long value) {
        byteBuffer.putLong(value);
        return this;
    }

    public MyByteBuffer addFloat(float value) {
        byteBuffer.putFloat(value);
        return this;
    }

    public MyByteBuffer addDouble(double value) {
        byteBuffer.putDouble(value);
        return this;
    }

}

File: cpp/src/my_byte_buffer.cc
#include "my_byte_buffer.h"

#include <cfloat>
#include <cstdio>
#include <cstdlib>
#include <cstring>

#include "MurmurHash3.h"

MyByteBuffer::MyByteBuffer() : position_(0) {
  byte_buffer_ = new char[128];
}

MyByteBuffer::MyByteBuffer(int capasity) : position_(0) {
  byte_buffer_ = new char[capasity];
}

MyByteBuffer::~MyByteBuffer() {
  if (byte_buffer_ != NULL) {
    delete byte_buffer_;
    byte_buffer_ = NULL;
  }
}

MyByteBuffer* MyByteBuffer::Clear() {
  position_ = 0;
  return this;
}

int MyByteBuffer::HashCode() {
  return HashCode(0);
}

int MyByteBuffer::HashCode(int seed) {
  int hash = 0;
  MurmurHash3_x86_32(byte_buffer_, position_, seed, &hash);
  return hash;
}

MyByteBuffer* MyByteBuffer::AddString(const std::string &value) {
  int len = value.length();
  memmove(byte_buffer_ + position_, value.c_str(), len);
  position_ += len;
  return this;
}

MyByteBuffer* MyByteBuffer::AddInt(int value) {
  int len = sizeof(value);
  memmove(byte_buffer_ + position_, &value, len);
  position_ += len;
  return this;
}

MyByteBuffer* MyByteBuffer::AddLong(int64_t value) {
  int len = sizeof(value);
  memmove(byte_buffer_ + position_, &value, len);
  position_ += len;
  return this;
}

MyByteBuffer* MyByteBuffer::AddFloat(float value) {
  int len = sizeof(value);
  memmove(byte_buffer_ + position_, &value, len);
  position_ += len;
  return this;
}

MyByteBuffer* MyByteBuffer::AddDouble(double value) {
  int len = sizeof(value);
  memmove(byte_buffer_ + position_, &value, len);
  position_ += len;
  return this;
}
